hw/top_earlgrey/syn/asic.constraints.sdc

```diff
@@ -162,6 +162,11 @@ set SPI_DEV_TCK 16.0
 create_clock -name SPI_DEV_CLK  -period ${SPI_DEV_TCK} [get_ports ${SPI_DEV_CLK_PIN}]
 set_clock_uncertainty ${SETUP_CLOCK_UNCERTAINTY} [get_clocks SPI_DEV_CLK]
 
+create_generated_clock -name SPI_DEV_IN_CLK -source SPI_DEV_CLK -edges {1 2 3} \
+    [get_pins top_earlgrey/u_spi_device/u_clk_spi_in_buf/clk_o]
+create_generated_clock -name SPI_DEV_OUT_CLK -invert -source SPI_DEV_CLK \
+    -edges {2 3 4} [get_pins top_earlgrey/u_spi_device/u_clk_spi_out_buf/clk_o]
```

Comment:
Out clock has negedge (inverted) clock to the SPI_CLK. As this clock is generated clock, `-waveform {}` cannot be an option. So I tried this. But I am not sure this is right solution or not.

Reply:
do we absolutely need to spec out the edges? because no matter which edge is used relatively inside the block, won't it always be timing checked for half cycle? Or are there situations where it checks both versions at the same time? one inverted and one not?

Reply:
If we can mandate the SPI_DEVICE timing always half clock period and assume SPI_DEV_IN_CLK / SPI_DEV_OUT_CLK have same phase then, I think it is also doable. Then,, the constraint should be..

```
create_generated_clock -name SPI_DEV_IN_CLK -source SPI_DEV_CLK -multiplied_by 2 [...]
create_generated_clock -name SPI_DEV_OUT_CLK -source SPI_DEV_CLK -multiplied_by 2 [...]
```

Reply:
sorry i think maybe i'm like a step behind, but what's the reason we need the multiplied by 2..? are we saying the internal clocks need to run faster than what is spec'd on the pads?

Reply:
I just meant creating fake clock that frequency is double to the SPI_CLK so that the datapath inside SPI_DEVICE should be met half clock period of SPI_CLK.

Reply:
o i think i see where you're going with this. Would it make sense though to just use the clock of the same frequency, and maybe constraint the incoming data to the negedge or something? The inside logic is always posedge right? I wonder if that means we should put a case analysis on the modes.

Reply:
Unfortunately, internal logic is also bounded by the half clock cycle. For example, in read command process block, the output data is read from sram in SPI_CLK (clk_spi_in_buf). The data goes out in clk_spi_out_buf. Also changing from Single IO mode to Dual or Quad IO mode signal originated in spi_in clock domain and ended in spi_out clock domain.

So, that's why I end up with `edges` to define the phase relationship between SPI_DEV_IN_CLK and SPI_DEV_OUT_CLK

Reply:
ah okay, so what i was trying to get at..

why not define SPI_DEV_OUT_CLK to be exactly the same as SPI_DEV_CLK, and then define SPI_DEV_IN_CLK to be the inverted version of SPI_DEV_CLK?

Once you have these two defined, and as long as they're not defined as async groups, wouldn't the tool automatically check between them and the half cycle paths?

This of course doesn't cover the opposite case, but we COULD define them too, but set those clocks to be async to the ones in the other set. 

Would that work? Just brainstorming here.

Reply:
That's what I achieved here. SPI_DEV_IN is in phase with SPI_DEV_CLK and SPI_DEV_OUT is 180 degree flipped to SPI_DEV_CLK. So, timing check is done in half SPI_CLK period.

Reply:
ah, but instead of using edges, why not use `-invert`, wouldn't that be more clear?

Reply:
DC tool was complaint that either `-multiplied_by`, `-divided_by`, or `-edges` should be in the option.. So, maybe adding -divided_by 1 and -invert?

Reply:
Revised and updated. PTAL